torchinductor/codegen/jiterator.py

```diff
@@ -0,0 +1,288 @@
+    torch.float32: "float",
+    torch.float64: "double",
+    torch.int64: "long",
+    torch.int32: "int",
+    torch.int16: "short",
+    torch.int8: "signed char",
+    torch.uint8: "unsigned char",
+    torch.bool: "bool",
+}
+INDEX_TYPE = "long"
+
+
+class CppPrinter(ExprPrinter):
+    def _print_ModularIndexing(self, expr):
+        x, div, mod = expr.args
+        x = self.paren(self.doprint(x))
+        div = self.paren(self.doprint(div))
+        mod = self.paren(self.doprint(mod))
+        if div != "1":
+            x = f"({x} / {div})"
+        return f"{x} % {mod}"
+
+    def _print_IndexingDiv(self, expr):
+        x, div = expr.args
+        x = self.paren(self.doprint(x))
+        div = self.paren(self.doprint(div))
+        return f"({x} / {div})"
+
+
+cexpr = CppPrinter().doprint
+class JiteratorOverrides(OpOverrides):
+    """Map element-wise ops to CUDA code"""
+
+    @staticmethod
+    def to_dtype(x, dtype):
+        return f"static_cast<{DTYPE_TO_CPP[dtype]}>({x})"
+
+    @staticmethod
+    def relu(x):
+        return f"{x} * ({x}>0)"
+
+    @staticmethod
+    def minimum(a, b):
+        return f"::min({a}, {b})"
+
+    @staticmethod
+    def maximum(a, b):
+        return f"::max({a}, {b})"
+
+    @staticmethod
+    def where(a, b, c):
+        return f"{a} ? {b} : {c}"
+
+    @staticmethod
+    def constant(val, dtype):
+        if val == float("inf"):
+            return f"std::numeric_limits<{DTYPE_TO_CPP[dtype]}>::infinity()"
+        elif val == float("-inf"):
+            return f"-std::numeric_limits<{DTYPE_TO_CPP[dtype]}>::infinity()"
+        return ops.to_dtype(repr(val), dtype)
+
+    @staticmethod
+    def index_expr(expr, dtype):
+        return ops.to_dtype(cexpr(V.kernel.rename_indexing(expr)), dtype)
+
+    @staticmethod
+    def masked(mask, body, other):
+        code = BracesBuffer()
+        var = V.kernel.cse.newvar()
+        assert isinstance(other, float)
+        if other == float("-inf"):
+            code.writeline(f"float {var} = -std::numeric_limits<float>::infinity();")
+        else:
+            assert False
+        code.writeline(f"if({mask})")
+        with V.kernel.swap_buffers(code), code.indent():
+            result = body()
+            code.writeline(f"{var} = {result};")
+        V.kernel.compute.splice(code)
+        return var
+
+    @staticmethod
+    def and_(a, b):
+        return f"{a} && {b}"
+
+    @staticmethod
+    def logical_not(a):
+        return f"!{a}"
+
+
+class JiteratorKernel(Kernel):
+    overrides = JiteratorOverrides
+    sexpr = cexpr
+    newvar_prefix = "auto "
+    suffix = ";"
+
+    def __init__(self, args):
+        super(JiteratorKernel, self).__init__(args)
+        self.call_ranges = None
+        self.ranges = None
+        self.itervars = None
+        self.reduction_depth = None
+        self.reduction_vars = {}
+
+    def load(self, name: str, index: sympy.Expr):
+        var = self.args.input(name)
+        index = self.rename_indexing(index)
```

Comment:
Here we aren't actually using the indexing formula.  So if this isn't a basic pointwise op things will fail.

As a short term fix you could add an assert that this is basic indexing by calling the indexer on the source node to see what a naive load looks like.

Longer term, I think the right fix is to rewrite the `index` into a call to `torch.as_strided()`

Something similar to:
https://github.com/jansel/torchdynamo/blob/345b44f3b6fee4e4b5e16a4dff1fad8e1c936b2d/torchinductor/codegen/triton.py#L281

First you can use some sympy `subs` calls to separate out the stride in each iteration axis:
```
offset = index.subs({v: 0 for v in iter_vars})
stride0 = index.subs({iter_vars[i]: 0 for i in range(len(iter_vars)) if i != 0}) - offset
stride1 = index.subs({iter_vars[i]: 0 for i in range(len(iter_vars)) if i != 1}) - offset
stride2 = index.subs({iter_vars[i]: 0 for i in range(len(iter_vars)) if i != 2}) - offset
...
```

Next, in the wrapper code (that calls into jiterator) you would do:
```
torch.as_strided(x, iteration_ranges, (stride0, stride1, stride2), offset)
```

This will work for everything except for indirect loads (from aten.index, embeddings, etc), and masked loads (from padding).

Note you will need to include the index in the hash key, since we might read the same buffer with multiple indexing formulas.  In this case we should pass them in as if they are different tensors.

Reply:
Jiterator handles indexing internally. 

Codegen just need to define the kernel for a single element, e.g 
```
code_string = "template <typename T> T inverted_where(bool cond, T a, T b){ return !cond ? a : b; }"
```

For more examples: see https://github.com/pytorch/pytorch/pull/77121

Reply:
I am aware of how Jiterator handles indexing.  `as_strided()` is the way to map from TorchInductor indexing to Jiterator indexing.

For an example try compiling:
```
x[::2] + x[1::2]
```

With your current code it will generate the incorrect result.

If you trace through the codegen you will see:
```
tmp0 = load("x", 2*sympy.Symbol("i0")) 
tmp1 = load("x", 2*sympy.Symbol("i0") + 1) 
tmp2 = ops.add(tmp0, tmp1)
```

Note a "normal" index to "x" would just be `sympy.Symbol("i0")` (without the `2*` or `+1`).

Since the current version just ignores the `index=...`, Jiterator won't do the correct thing.  You need to do the `as_strided()` thing to convey the correct indexing formula to Jiterator.